File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.kt
class Solution {
    fun largestRectangleArea(heights: IntArray): Int {
        var queue = mutableListOf<IntArray>()
        var output = 0
        for (i in 0 until heights.count()) {
            val currentHeight = heights[i]
            var currentIndex = i
            while (queue.count() > 0 && queue.last()[1] > currentHeight) {
                val last = queue.removeLast()
                val size = i - last[0]
                val square = size * last[1]
                output = maxOf(output, square)
                currentIndex = last[0]
            }
            queue.add(intArrayOf(currentIndex, currentHeight))

        }
        for (i in 0 until queue.count()) {
            output = maxOf(output, (heights.count() - queue[i][0]) * queue[i][1])
        }
        return output
    }
}

File: Data Structures & Algorithms/merge-intervals/submission-3.swift
class Solution {
    func merge(_ intervals: [[Int]]) -> [[Int]] {
        var sorted = intervals.sorted { a, b in 
            if a[0] == b[0] {
                a[1] < b[1]
            } else {
                a[0] < b[0]
            }
        }

        var output = [[Int]]()
        output.append(sorted[0])

        for i in 1 ..< sorted.count {
            var last = output[output.count - 1]
            var current = sorted[i]
            if last[1] >= current[0] {
                output.removeLast()
                output.append([min(last[0], current[0]), max(last[1], current[1])])
            } else {
                output.append(current)
            }
        }

        return output

    }
}
